symforce/slam/imu_preintegration/preintegrated_imu_measurements.h

```diff
@@ -0,0 +1,96 @@
+/* ----------------------------------------------------------------------------
+ * SymForce - Copyright 2022, Skydio, Inc.
+ * This source code is under the Apache 2.0 license found in the LICENSE file.
+ * ---------------------------------------------------------------------------- */
+
+#pragma once
+
+#include <Eigen/Dense>
+
+#include <sym/rot3.h>
+#include <sym/util/epsilon.h>
+
+namespace sym {
+
+/**
+ * Struct of Preintegrated IMU Measurements (not include the covariance of change in orientation,
+ * velocity, and position).
+ */
+template <typename Scalar>
+struct PIMNoCovariance {
+  using Vector3 = Eigen::Matrix<Scalar, 3, 1>;
+  using Matrix3 = Eigen::Matrix<Scalar, 3, 3>;
+
+  sym::Rot3<Scalar> DR;
+  Vector3 Dv;
+  Vector3 Dp;
+  Matrix3 DR_D_gyro_bias;
+  Matrix3 Dv_D_gyro_bias;
+  Matrix3 Dv_D_accel_bias;
+  Matrix3 Dp_D_gyro_bias;
+  Matrix3 Dp_D_accel_bias;
+  Vector3 accel_bias;
+  Vector3 gyro_bias;
+
+  double integrated_dt;
+
+  /**
+   * Initialize instance struct with accel_bias and gyro_bias and all other values
+   * zeroed out (scalars, vectors, and matrices) or set to the identity (DR).
+   */
+  PIMNoCovariance(const Vector3& accel_bias, const Vector3& gyro_bias);
+};
+
+/**
+ * Struct of preintegrated IMU measurements (using on manifold pre-integration).
+ */
+template <typename Scalar>
+struct PreintegratedImuMeasurements {
+  using Vector3 = typename PIMNoCovariance<Scalar>::Vector3;
+  using Matrix3 = typename PIMNoCovariance<Scalar>::Matrix3;
+  using Matrix9 = Eigen::Matrix<Scalar, 9, 9>;
+
+  PIMNoCovariance<Scalar> pim;
+  Matrix9 covariance;  // covariance of [DR.ToTangent(), Dv, Dp]
+
+  /**
+   * Initialize with given accel_bias and gyro_bias
+   */
+  PreintegratedImuMeasurements(const Vector3& accel_bias, const Vector3& gyro_bias);
+
+  /**
+   * Integrate a new measurement.
+   *
+   * Precondition:
+   *   *this was initialized by a constructor and has only been mutated by IntegrateMeasurement
+   *     (called in order on consecutive IMU measurements)
+   *   measured_accel is the next accelerometer measurement after the last integrated measurement
+   *   measured_gyro is the next gyroscope measurement after the last integrated measurement
+   *   accel_cov is the covariance of the accelerometer measurement
+   *   gyro_cov is the covariance of the gyroscope measurement
+   *   dt is the time span over which these measurements were made
+   *
+   * Postcondition:
+   *   If the orientation, velocity, and position were R0, v0, and p0 at the start of the first
+   *   integrated IMU measurements, and Rf, vf, and pf are the corresponding values at the end of
+   *   the measurement described by the arguments, DT is the total time covered by the integrated
+   *   measurements, and g is the vector of acceleration due to gravity, then
+   *   - pim.DR -> R0.inverse() * Rf
```

Comment:
Suggest we follow the convention in the paper and use subscripts `i` and `j` to denote the first and last timestamp.

Reply:
I considered that, but I was worried that without the paper for context it would make less sense. In the paper, `i` and `j` were just two example indices among many (so it made sense to use those indices) (but even then, on occasion I had to think for a moment whether or not index `i` was the earlier time step or the other way around).

But here there's no reason to reference other measurements, so absent the paper, there's no reason to use `i` and `j`.